pkg/sdn/plugin/node.go

```diff
@@ -127,11 +139,138 @@ func (node *OsdnNode) Start() error {
+	}
+	go utilwait.Forever(node.watchCni, 0)
+
 	node.markPodNetworkReady()
 
 	return nil
 }
 
+type podWorker struct {
+	node     *OsdnNode
+	requests []*podRequest
+	lock     sync.Mutex
+}
+
+func newPodWorker(node *OsdnNode) *podWorker {
+	return &podWorker{
+		node:     node,
+		requests: make([]*podRequest, 0),
+	}
+}
+
+func (w *podWorker) addRequest(request *podRequest) {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.requests = append(w.requests, request)
+	if len(w.requests) == 1 {
+		// Start proessing when the first request is added
+		go w.process()
+	}
+}
+
+func (w *podWorker) numRequests() int {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	return len(w.requests)
+}
+
+func (w *podWorker) process() {
+	for {
+		// Grab an item off the queue
+		w.lock.Lock()
+		if len(w.requests) == 0 {
+			w.lock.Unlock()
+			return
+		}
+		request := w.requests[0]
+		w.requests = w.requests[1:]
+		w.lock.Unlock()
+
+		log.V(5).Infof("Handling pod network request %#v", request)
+
+		var err error
+		result := &podResult{code: http.StatusOK}
+		switch request.command {
+		case CNI_ADD:
+			var ipamResult *cnitypes.Result
+			ipamResult, err = w.node.setupPod(request)
+			if err == nil {
+				result.responseJson, err = json.Marshal(ipamResult)
+			}
+		case CNI_DEL:
+			err = w.node.teardownPod(request)
+		case CNI_UPDATE:
+			err = w.node.updatePod(request)
+		}
+
+		if err != nil {
+			result.code = http.StatusBadRequest
+			result.err = fmt.Sprintf("%v", err)
+		}
+		log.V(5).Infof("Returning pod network result %#v", result)
+		request.result <- result
+	}
+}
+
+func (node *OsdnNode) watchCni() {
+	for {
+		select {
+		case request := <-node.podRequestChan:
+			log.V(5).Infof("Dispatching pod network request %#v", request)
+
+			podKey := fmt.Sprintf("%s/%s", request.podNamespace, request.podName)
+			node.cniLock.Lock()
+			worker, ok := node.cniWorkers[podKey]
+			if !ok {
+				worker = newPodWorker(node)
+				node.cniWorkers[podKey] = worker
+			}
+			node.cniLock.Unlock()
+
+			// Add the request to the worker
+			log.V(5).Infof("Added pod network request %#v", request)
+			worker.addRequest(request)
+
+			// Clean up finished workers
+			node.cniLock.Lock()
+			for key, worker := range node.cniWorkers {
+				// Safe to compare to 0 after unlocking as this
+				// is the only goroutine that adds things to the
+				// workers queue
+				if worker.numRequests() == 0 {
+					delete(node.cniWorkers, key)
+				}
+			}
+			node.cniLock.Unlock()
+		}
+	}
+}
+
+// FIXME: this should eventually go into kubelet via a CNI UPDATE/CHANGE action
+// See https://github.com/containernetworking/cni/issues/89
+func (node *OsdnNode) UpdatePod(namespace string, name string, id kubeletTypes.ContainerID) error {
+	resultChan := make(chan *podResult)
+	node.podRequestChan <- &podRequest{
+		command:      CNI_UPDATE,
+		podNamespace: namespace,
+		podName:      name,
+		containerId:  id.String(),
+		netns:        "/blah/something", // plugin doesn't care about namespace
```

Comment:
Comment doesn't appear to be true; pod.go:updatePod() calls `getVethInfo(req.netns, "eth0")` which looks like it will return an error if netns is bogus.